src/Catalog/Helpers/GalleryDatabaseQueryService.cs

```diff
@@ -178,7 +179,7 @@ internal static string BuildDb2CatalogSqlQuery(Db2CatalogCursor cursor)
             return $@"SELECT TOP {cursor.Top} WITH TIES
```

Comment:
won't this `TOP` value be a little weird now? Like if TOP is 20 today and one package has 10 vulnerabilities, won't this only return 11 unique packages now (but still 20 rows)?

Reply:
BTW I might be totally wrong, I have not used `WITH TIES` much

Reply:
It's certainly something I should investigate.

Reply:
@joelverhagen Here's how TIES seems to operate--SELECT TOP 550 (this is the constant we use--MaxPageSize) will give the top 550 rows based on ordering. WITH TIES will add rows after the 550th row whose ordinal columns equal those of row 550. So, your concern about the 10-vulnerability scenario is founded--you would only get 541 package versions in that case (we don't actually have a situation yet where a  package version has >1 vulnerabilities but we should design for it).

Prior to this change we were ordering only by LastEdited, which means TIES would be rows after row n (e.g. n=550) whose LastEdited value equals that of row n.

This change introduces the package key into the ordering, which means that if we have:
row 549:     LastEdited: '2020-06-30 00:30:34.986', Key: 101, Vulnerability: 200
row 550:     LastEdited: '2020-06-30 00:30:34.986', Key: 101, Vulnerability: 453
row 551:     LastEdited: '2020-06-30 00:30:34.986', Key: 101, Vulnerability: 664
...without WITH TIES we run the risk of splitting a single package version catalog entry into two, if this series is broken at max page size n like above. WITH TIES will include row 551 in our set. When we build our catalog entry, of course, these 3 rows become a single json blob.

Really what we have here is a decision to make:
Option A: MaxPageSize will continue to operate as a maximum, and for every package version in the dataset which has i>1 vulnerabilities, our page size will diminish by i-1 entries. We could decide that this is good enough, allowing MaxPageSize to be our safety net but to no longer create uniform page sizes.

Option B: If we do want MaxPageSize to act as an exact requirement on page size (except of course when there are fewer available), our query here may need two passes:
- the first would use TOP, grouping all vulnerability key values, giving us a dataset containing only 1 row per package version (unique on package.key). We'd store this dataset in a temporary table
- the second being a >cursor only (no TOP) and inner joining with the temp table to give us our final set.
Another possibility is we do this as a single query, grouping again on package.key and aggregating vulnerabilty data into delimited strings which we then process. This seems the better idea, actually.

So, I'm all ears on opinions re this decision. I personally would expect that option A is the better course of action. What do you think?

Reply:
/cc @agr @ryuyu ^^^

Reply:
@joelverhagen Have taken another look and our cursor size is 20 (from the config). I'm going to revisit this in a bit. I may go for the aggregated string version of option B above.

Reply:
After some perf analysis of 
1. GROUP BY all fields in original query to aggregate vulnerability fields into delimited strings for later processing (giving us 1 row per package), 
2. GROUP BY key field only, taking MAX of each original field and aggregating vulnerabilty fields as in 1.,
3. moving existing query into an inner query to preserve TOP count, and then joining to vulnerability tables and utilising existing control break logic to process additional rows,
...3 is the clear winner, so I have implemented this approach.